# Exploring HackerNews Posts 
Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit.
Our dataset has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions

Dataset includes the following entries :
id: The unique identifier from Hacker News for the post
title: The title of the post
url: The URL that the posts links to, if it the post has a URL
num_points: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
num_comments: The number of comments that were made on the post
author: The username of the person who submitted the post
created_at: The date and time at which the post was submitted


We are interested in posts whose titles begin with either Ask HN or Show HN.
Ask HN : Ask the HN community questions like  How to improve my personal website?
Show HN : Users submit posts to show HN community a product,a project developed by them , something interesting etc


Our objective here is to compare these two types of posts to determine the following:

Q1. Do Ask HN or Show HN receive more comments on average?
Q2. Do posts created at a certain time receive more comments on average?

In [1]:
from csv import reader 
open_file = open('hacker_news.csv')
read_file = reader(open_file)
hn = list(read_file)
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


#### Extract the first row 

In [2]:
headers = hn[0]
print(headers)
hn = hn[1:]
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


#### Extracting the posts beginning with Ask HN or Show HN 

In [3]:
ask_posts = []
show_posts = []
other_posts = []
for row in hn:
    title = row[1]
    title_lower = title.lower()
    if title_lower.startswith('ask hn'):
        ask_posts.append(row)
    elif title_lower.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

In [4]:
total_ask_comments = 0
for post in ask_posts:
    num_comments = int(post[4])
    total_ask_comments += num_comments
    
avg_ask_comments = total_ask_comments/len(ask_posts)
print(avg_ask_comments)

14.038417431192661


#### Find out avg comments in show HN posts

In [5]:
total_show_comments = 0
for post in show_posts:
    num_comments = int(post[4])
    total_show_comments += num_comments

avg_show_comments = total_show_comments/len(show_posts)
print(avg_show_comments)
    

10.31669535283993


#### Finding 
Between show posts and ask posts - ask posts receive more comments on average which is approximately 14 whereas show posts receive approximately 10 comments .So we'll focus our remaining analysis just on these posts .

#### New Observations 
TO calculate amount of ask posts created in each hour of day along with number of comments 
To calculate the average number of comments ask posts receive by hour created.

In [8]:
import datetime as dt
result_list =[]
for post in ask_posts:
    created_at = post[6]
    num_comments = int(post[4])
    result_list.append([created_at,num_comments])

counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    date = row[0]
    comment = row[1]
    date_obj = dt.datetime.strptime(date,"%m/%d/%Y %H:%M")
    hour = date_obj.strftime("%H") 
    
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comment
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comment



 To create a list of lists containing the hours during which posts were created and the average number of comments those posts received.
 
 counts by hour = no of posts per hour
 comments by hour = no of comment in that hour 
 

In [9]:
avg_by_hour = []
for hour in comments_by_hour:
    avg_by_hour.append([hour,comments_by_hour[hour]/counts_by_hour[hour]])
    
print(avg_by_hour)

[['06', 9.022727272727273], ['12', 9.41095890410959], ['15', 38.5948275862069], ['02', 23.810344827586206], ['14', 13.233644859813085], ['08', 10.25], ['03', 7.796296296296297], ['22', 6.746478873239437], ['23', 7.985294117647059], ['10', 13.440677966101696], ['19', 10.8], ['00', 8.127272727272727], ['07', 7.852941176470588], ['05', 10.08695652173913], ['17', 11.46], ['20', 21.525], ['18', 13.20183486238532], ['21', 16.009174311926607], ['04', 7.170212765957447], ['13', 14.741176470588234], ['09', 5.5777777777777775], ['01', 11.383333333333333], ['16', 16.796296296296298], ['11', 11.051724137931034]]


#### Sorting the list of lists 
Print the five highest values 

In [10]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])
print(swap_avg_by_hour)

[[9.022727272727273, '06'], [9.41095890410959, '12'], [38.5948275862069, '15'], [23.810344827586206, '02'], [13.233644859813085, '14'], [10.25, '08'], [7.796296296296297, '03'], [6.746478873239437, '22'], [7.985294117647059, '23'], [13.440677966101696, '10'], [10.8, '19'], [8.127272727272727, '00'], [7.852941176470588, '07'], [10.08695652173913, '05'], [11.46, '17'], [21.525, '20'], [13.20183486238532, '18'], [16.009174311926607, '21'], [7.170212765957447, '04'], [14.741176470588234, '13'], [5.5777777777777775, '09'], [11.383333333333333, '01'], [16.796296296296298, '16'], [11.051724137931034, '11']]


In [16]:
sorted_swap = sorted(swap_avg_by_hour, reverse = True)
print("Top 5 Hours for Ask Posts Comments")
for avg, hr in sorted_swap[:5]:
    print("{}: {:.2f} average comments per post".format(dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg))

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


## Conclusion
The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.
According to the data set documentation, the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.
